core/src/main/kotlin/fr/sncf/osrd/stdcm/graph/VisitedNodes.kt

```diff
@@ -28,86 +25,151 @@ import kotlin.Double.Companion.POSITIVE_INFINITY
+ * criteria to chose one path over another, we likely need to add an extra visited range type*.
+ * Otherwise, we may consider that a range is visited despite being better according to the new
+ * criteria.
  */
 data class VisitedNodes(val minDelay: Double) {
 
     /** Data class representing a space location. Must be usable as map key. */
     data class Fingerprint(
         val identifier: EdgeIdentifier,
         val waypointIndex: Int,
-        val startOffset: Distance
+        val startOffset: Distance,
     )
 
-    /** Parameters for either function, having a class avoids repetitions */
+    /**
+     * Parameters for either function (is visited / mark as visited), having a class avoids
+     * repetitions
+     */
     data class Parameters(
         var fingerprint: Fingerprint?,
-        val startTime: Double,
-        val duration: Double,
+        val timeData: TimeData,
         val maxMarginDuration: Double,
-        val baseNodeCost: Double,
-    )
+        val remainingTimeEstimation: Double = 0.0,
+    ) {
+        val nodeCost = timeData.totalRunningTime + remainingTimeEstimation
+    }
+
+    /** Any class that implements this interface may be added to the visited ranges. */
+    private sealed interface ConditionallyVisitedRange {
+        /** Is the given (sub)range is covered by `this`? */
+        fun isVisited(
+            range: Range<Double>,
+            parameters: Parameters,
+        ): Boolean
+
+        /**
+         * Returns the range that is most likely to define elements as "already visited". Used to
+         * merge overlapping ranges in the range map.
+         */
+        fun mergeWith(other: ConditionallyVisitedRange?): ConditionallyVisitedRange
+    }
 
     /**
-     * Linear function with slope of 1. Used to represent the cost to reach a given time with margin
+     * The given range is already visited by changing departure time. Such ranges are *always*
+     * already visited.
      */
-    data class LinearFunction(private val y0: Double) : Comparable<LinearFunction> {
-        fun apply(x: Double): Double = y0 + x
+    private data object VisitedWithDepartureTimeChange : ConditionallyVisitedRange {
+        override fun isVisited(
+            range: Range<Double>,
+            parameters: Parameters,
+        ): Boolean {
+            // If the other option were better, it would have been visited first
+            return true
+        }
 
-        override fun compareTo(other: LinearFunction): Int {
-            return y0.compareTo(other.y0)
+        override fun mergeWith(other: ConditionallyVisitedRange?): ConditionallyVisitedRange {
+            return this
         }
     }
 
     /**
-     * For each location, the time ranges that have been visited. We know they can always be
-     * ignored.
+     * The given range is already visited by adding stop durations. The other candidates are visited
+     * if they add more stop time than this.
```

Comment:
The difference with the next class is that the other candidates are only visited if their cost is lower. Might be good to have it written in the main class description (i know you also wrote it a bit further down).

Reply:
Fixed in 0c57be6390f29f51b60b1078138c60211d4dfd1f (new phrasing)